modules/commons/src/test/java/org/apache/synapse/commons/datasource/DataSourceInformationRepositoryTest.java

```diff
@@ -0,0 +1,195 @@
+    private static final String ALIAS = "testAlias";
+    private static final String ERR_DATASOURCE = "notExistingDataSourceName";
+
+    /**
+     * Initializing reflection variables.
+     *
+     * @throws NoSuchFieldException
+     */
+    @BeforeClass
+    public static void initialize() throws NoSuchFieldException {
+        information.setAlias(ALIAS);
+        repositoryListener = DataSourceInformationRepository.class.getDeclaredField("listener");
+        repositoryListener.setAccessible(true);
+        dataSourceInformationMap = DataSourceInformationRepository.class.getDeclaredField("dataSourceInformationMap");
+        dataSourceInformationMap.setAccessible(true);
+        secretResolver = DataSourceInformationRepository.class.getDeclaredField("secretResolver");
+        secretResolver.setAccessible(true);
+    }
+
+    /**
+     * Test set, get and remove methods of repository listener.
+     */
+    @Test
+    public void testSetGetRemoveRepositoryListener() throws IllegalAccessException {
+        DataSourceInformationRepositoryListener listener = new DataSourceRepositoryManager
+                (inMemoryDataSourceRepository, jndiBasedDataSourceRepository);
+        repository.setRepositoryListener(listener);
+        Assert.assertNotNull("new repository listener should be added", repositoryListener.get(repository));
+        try {
+            repository.setRepositoryListener(listener);
+            Assert.fail("execution successful without expected exception");
+        } catch (Exception ex) {
+            Assert.assertEquals("asserting exception message when add an existing lister again",
+                    "There is already a DataSourceInformationRepositoryListener associated with " +
+                            "'DataSourceInformationRepository", ex.getMessage());
+            Assert.assertEquals("asserting exception when add an existing lister again", SynapseCommonsException
+                    .class, ex.getClass());
+        }
+        DataSourceInformationRepositoryListener tempListener = repository.getRepositoryListener();
+        Assert.assertEquals("received listener should be same as the inserted one", listener, tempListener);
+        repository.removeRepositoryListener();
+        Assert.assertNull("listener should be removed", repositoryListener.get(repository));
+    }
+
+    /**
+     * Test setRepositoryListener with null parameter.
+     */
+    @Test
+    public void testSetRepositoryListenerNull() {
+        try {
+            repository.setRepositoryListener(null);
+            Assert.fail("execution successful without expected exception");
+        } catch (Exception ex) {
+            Assert.assertEquals("asserting exception message", "Provided " +
+                    "DataSourceInformationRepositoryListener instance is null", ex.getMessage());
+            Assert.assertEquals("asserting exception class", SynapseCommonsException.class, ex.getClass());
```

Comment:
Can't we use expected=SynapseCommonsException.class to test thrown exception? Or expected exception rule?

Reply:
+1 , better to use expectedException

Reply:
expected=SynapseCommonsException.class does not give us a way to assert exception messages.
expected exception rule works same as try catch, except in try catch we have flexibility to use various assert methods
So, do I have to make this change?

Reply:
Since exception message is also asserted try/catch should be fine. However JUnit also provides  @Rule annotation where you can assert for exception type + message. Since you are checking for 'SynapseCommonsException' mostly, below approach might be cleaner. 
@Rule
    public SynapseCommonsException thrown = SynapseCommonsException.none();
Then you could add in your test-cases like,
thrown.expect();
thrown.expectMessage();